staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -768,6 +770,131 @@ func (r *Request) Watch(ctx context.Context) (watch.Interface, error) {
 	}
 }
 
+type WatchListResult struct {
+	// err holds any errors we might have received
+	// during streaming.
+	err error
+
+	// items hold the collected data
+	items []runtime.Object
+
+	// initialEventsEndBookmarkRV holds the resource version
+	// extracted from the bookmark event that marks
+	// the end of the stream.
+	initialEventsEndBookmarkRV string
+
+	// gv represents the API version
+	// it is used to construct the final list response
+	// normally this information is filled by the server
+	gv schema.GroupVersion
+}
+
+func (r WatchListResult) Into(obj runtime.Object) error {
+	if r.err != nil {
+		return r.err
+	}
+
+	listPtr, err := meta.GetItemsPtr(obj)
+	if err != nil {
+		return err
+	}
+	listVal, err := conversion.EnforcePtr(listPtr)
+	if err != nil {
+		return err
+	}
+	if listVal.Kind() != reflect.Slice {
+		return fmt.Errorf("need a pointer to slice, got %v", listVal.Kind())
+	}
+
+	if len(r.items) == 0 {
+		listVal.Set(reflect.MakeSlice(listVal.Type(), 0, 0))
+	} else {
+		listVal.Set(reflect.MakeSlice(listVal.Type(), len(r.items), len(r.items)))
+		for i, o := range r.items {
+			listVal.Index(i).Set(reflect.ValueOf(o).Elem())
+		}
+	}
+
+	listMeta, err := meta.ListAccessor(obj)
+	if err != nil {
+		return err
+	}
+	listMeta.SetResourceVersion(r.initialEventsEndBookmarkRV)
+
+	typeMeta, err := meta.TypeAccessor(obj)
+	if err != nil {
+		return err
+	}
+	version := r.gv.Version
+	if len(r.gv.Group) > 0 {
+		version = fmt.Sprintf("%s/%s", r.gv.Group, r.gv.Version)
+	}
+	typeMeta.SetAPIVersion(version)
+	typeMeta.SetKind(reflect.TypeOf(obj).Elem().Name())
+
+	return nil
+}
+
+// WatchList establishes a stream to get a consistent snapshot of data
+// from the server as described in https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list#proposal
+//
+// Note that the watchlist requires properly setting the ListOptions
+// otherwise it just establishes a regular watch with the server.
+// Check the documentation https://kubernetes.io/docs/reference/using-api/api-concepts/#streaming-lists
+// to see what parameters are currently required.
+func (r *Request) WatchList(ctx context.Context) WatchListResult {
+	w, err := r.Watch(ctx)
```

Comment:
Do we want to return an error when the client-go flag hasn't been set ? 
Alternatively we could document that this is expected from the higher layers.

Reply:
> Also - we should return an error if the flags for watch will not result in WatchList call on the server side (i.e. if SendInitialEvents and/or EnableWatchBookmarks are not set).

I wanted to avoid checking the parameters so that we can update the server without updating the client. Also we don't validate params for other types of request, i.e `list`

Reply:
I'm not saying we need full validation here - what I'm saying is that for me as a user of this method, I don't want to understand that I need to set 3 different params - I want this to be handled for me.

So I probably even mean something different - shouldn't we handle setting appropriate parameters for the user?

Reply:
The rest client is a low level component. Usually, people use the generated client. As far as I can tell, parameters are set by higher-level layers, like the generated client and not the rest client. For example https://github.com/kubernetes/kubernetes/blob/7a37df6d1ab8926a95574f0aab3a9978d20d4e35/staging/src/k8s.io/client-go/kubernetes/typed/apps/v1/deployment.go#L109

I think we should do the same.

Reply:
we chatted about that offline - given that don't set anything in other methods, we should probably not touch it here for consistency either.

I still think that some simple check (e.g. checking if InitialEvents is set) would be useful here to fail fast.
[Similarly as it would be useful to check if watch is set inside Watch() method]. But I think it may be a follow-up commit/PR where we can discuss it.

Reply:
added a TODO